storage/types.go

```diff
@@ -33,19 +33,61 @@ type Node struct {
 }
 
 // NodeID uniquely identifies a Node within a versioned MerkleTree.
+// NodeIDs as presented to the storage layer will always have a multiple
+// of 8 bits as they identify the root of a 256 element subtree.
+//
+// Reading paths right to left is the natural order when traversing from
+// leaves towards the root. However, for internal nodes the rightmost bits
+// of the IDs are not aligned on a byte boundary so care must be taken.
+//
+// Note that some of the APIs count bits with 0 being the rightmost
+// in the entire path array when some of these might not be significant.
 type NodeID struct {
-	// path is effectively a BigEndian bit set, with path[0] being the MSB
+	// Path is effectively a BigEndian bit set, with Path[0] being the MSB
 	// (identifying the root child), and successive bits identifying the lower
 	// level children down to the leaf.
 	Path []byte
 	// PrefixLenBits is the number of MSB in Path which are considered part of
 	// this NodeID.
 	//
 	// e.g. if Path contains two bytes, and PrefixLenBits is 9, then the 8 bits
-	// in Path[0] are included, along with the lowest bit of Path[1]
+	// in Path[0] are included, along with the MSB of Path[1]. The remaining
+	// 7 bits are not significant.
 	PrefixLenBits int
 }
 
+// Some Additional NodeID examples and Documentation.
+//
+// Consider a hypothetical case of a tree of size 256 and max path length 8.
+// The IDs within this tree will at most have 8 significant bits. To specify
+// the leaf with index 95 requires 8 bits to match the tree height so the
+// Path array would be: [0x5f] with PrefixLength 8. Moving up the tree two
+// levels from this node the depth would be 2 and index 23. This gives a Path
+// array of [0x5c] and PrefixLength is 6, which means the lowest 2 bits are not
+// significant. In binary 0x5c is: 0b01011100 and 23 decimal is: 0b000010111
+// so the index has been shifted left by 2 places in the NodeID. Reading
+// from left to right after 6 bits the node has been reached. The remaining
+// two bits therefore don't matter.
+//
+// Larger tree sizes as used by Logs and Maps work in exactly the same
+// way but with the Path taking up multiple bytes. For example if we move
+// to a tree size of up to 65535 there will be two Path bytes. For a leaf
+// in this tree at depth = 0, index = 19 the resulting NodeID has a
```

Comment:
To me the "depth" word seems counterintuitive, I imagine it being the distance from the root rather than the distance from leaves. IMO "level" describes the thing better, but that might be my English.

Reply:
We've used both depth and level in many places. In some cases unhelpfully. I'm not trying to fix it in this PR as it's too big a job.

Reply:
I'm using 'depth' here as it's the name of the parameter passed to `NewNodeIDWithPrefix`.

Reply:
Also, [this](https://stackoverflow.com/questions/2603692/what-is-the-difference-between-tree-depth-and-height).

How about `TODO(...): Use "level" instead of "depth" consistently.`, and making this PR the first step towards it (the new comments at least)?

Reply:
It affects more than this file though. I've already spent more time on this and related activities than I intended.